src/main.rs

```diff
@@ -18,26 +18,167 @@
 //!         ),
 //!     ]
 //! ```
+
 #![deny(missing_docs)]
 #![deny(clippy::all)]
+//#![feature(proc_macro_hygiene, decl_macro)]
 
 mod bundle;
 mod config;
 mod virtfs;
 mod workload;
 
+#[macro_use]
+extern crate serde_derive;
+
+use openssl::asn1::Asn1Time;
+use openssl::hash::MessageDigest;
+use openssl::pkey::PKey;
+use openssl::rsa::Rsa;
+use std::net::{IpAddr, SocketAddr};
+use std::path::Path;
+use warp::Filter;
+#[derive(Serialize, Deserialize)]
+struct Payload {
+    encoding: String,
+    contents: Vec<u8>,
+}
+
 use log::info;
+/// Source of the key to use for TLS
+//pub const KEY_SOURCE: &str = "file-system";
+pub const KEY_SOURCE: &str = "generate";
 
-fn main() {
-    let _ = env_logger::try_init_from_env(env_logger::Env::default());
+#[tokio::main]
+async fn main() {
+    //This required when calling from Rust std::process::command.  Recorded
+    // to allow debugging.
+    //    let args: Vec<String> = std::env::args().skip(1).collect();
+    let args: Vec<String> = std::env::args().collect();
+    let listen_address: &str = &args[0];
+    let listen_port: &str = &args[1];
+    let listen_socketaddr = SocketAddr::new(
+        listen_address.parse::<IpAddr>().unwrap(),
+        listen_port.parse().unwrap(),
+    );
+    let (server_key, server_cert) = get_credentials_bytes(listen_address);
 
-    let mut args = std::env::args().skip(1);
-    let path = args.next().unwrap();
-    let vars = std::env::vars();
+    // POST /payload
+    let workload = warp::post()
+        .and(warp::path("payload"))
+        .and(warp::body::json())
+        .and_then(payload_launch);
 
-    let bytes = std::fs::read(&path).expect("Unable to open file");
+    let routes = workload;
+    warp::serve(routes)
+        .tls()
+        .cert(&server_cert)
+        .key(&server_key)
+        .run(listen_socketaddr)
+        .await;
+}
 
-    let result = workload::run(&bytes, args, vars).expect("Failed to run workload");
+fn create_new_runtime(recvd_data: &[u8]) {
+    println!("About to attempt new runtime creation");
+    let _ = env_logger::try_init_from_env(env_logger::Env::default());
+    //TODO - get args these from main() if required
+    //    let args = std::env::args().skip(1);
+    let dummy_arr: [&str; 1] = [""];
+    let vars = std::env::vars();
 
+    let result = workload::run(recvd_data, &dummy_arr, vars).expect("Failed to run workload");
+    println!("Got result (println) {:#?}", result);
     info!("got result: {:#?}", result);
 }
+
+async fn payload_launch(payload: Payload) -> Result<impl warp::Reply, warp::Rejection> {
+    format!("Received a {} file", payload.encoding);
+    println!("Received a {} file", payload.encoding);
+    create_new_runtime(&payload.contents);
+    Ok(warp::reply::with_status(
+        "Payload received",
+        warp::http::StatusCode::OK,
+    ))
```

Comment:
I think the flow of this function is wrong. If I'm reading it correctly, it won't respond to the client until the wasm code has finished running.

Reply:
It is still wrong even in the single threaded case.

We basically want one set of APIs for adding wasm modules to the Keep and another API to start execution. Adding a module to the keep should parse the modules and make sure they are valid before returning `OK`. OTOH, the `go()` API should immediately return `OK`, tear down all further (external) communication and then start wasm execution.